chapter/related-work.md

```diff
@@ -36,10 +36,96 @@ Additionally, this part aims to recognize alternative approaches to the LSP, in
+   Undefined capabilities cannot be added to the context menu.
+
+   In the case of CodeCompass the developers made up for this by using the code completion feature as an alternative dynamic context menu.
+
+2. > "LSP does not support displaying pictures (diagrams)".
+
+   CodeCompass generates diagrams for selected code.
+   Yet, there is no image transfer included with the LSP.
+   Since the LSP is based on JSON-RPC messages, the authors' solution was to define a new command, specifically designed to tackle this non-standard use case.
+
+
+Missing features of the protocol such as the ones pointed out by Mészáros et al. appear frequently, especially in complex language servers or ones that implement more than basic code processing.
+
+The rust-analyzer defines almost thirty non-standard commands [@rust-analyzer-extensions], to enable different language specific actions.
+
+Taking the idea of the CodeCompass project further, Rodriguez-Echeverria et al. propose a generic extension of the LSP for graphical modeling [@lsp-for-graphical-modeling].
+Their approach is based on a generic intermediate representation of graphs which can be generated by language servers and turned into a graphical representation by the client implementation.
+
+Similarly, in [@Specification-Language-Server-Protocol,@decoupling-core-analysis-support] the authors describe a method to develop language agnostic LSP extensions.
+In their work they defined a language server protocol for specification languages (SLSP) which builds on top of the existing LSP, but adds several additional commands.
+The commands are grouped by their use case in the domain of specification languages and handled by separate modules of the client extension implementing support for the SLSP.
+Following the LSP example and defining commands language agnostic instead of tied to a specific language, allows to maintain the initial purpose of the LSP.
+Since the extensions can be incorporated by specific implementations of language servers in the same domain, a single client implementation serves multiple languages.
+The authors point out that while their approach specializes in specification languages, the idea can be transferred to other areas.
+
+### Language Server Index Format
+
+The Language Server Index Format, or short LSIF, is an augmentation of the LSP.
+Since the LSP requires a language server to actively analyze files and answer requests, its use is typically constrained to local installations.
+As pointed out in the introducing blog post [@lsif-blog-post], several use cases exist where the LSP approach fails due to resource limits.
+The article explicitly names web based platforms such as GitHub[@github] or Sourcegraph[@sourcegraph].
+
+The LSIF aims to provide the features of the LSP without the need of actively running a language server.
+Instead, "language servers or other programming tools to emit their knowledge about a code workspace" as a LSIF compliant JSON report.
```

Comment:
Yep, better. Here is another suggestion, to clarify a few things (e.g. static code -> you can't edit the code online), feel free to use/modify/merge with your suggestion/drop.

```suggestion
Nowadays, code hosting platforms are an integral part of the developer toolset (GitHub[@github], Sourcegraph[@sourcegraph], GitLab, Sourceforge, etc.). Those platforms commonly displays code simply as text, highlighted at best.
LSP-like features would make for a great improvement for code navigation and code reading online, but analysis would be very wasteful if it needed to be done each time a visitor load a chunk of code. Since the hosted code is most often not editable and precisely versioned, code analysis could be performed ahead of time, once and for all for each version.
The LSIF (Language Server index Format) specifies a schema for the output of such ahead of time code analysis.
Clients can then provide efficient code intelligence using LSIF data provided in one shot by the server.
```